### ISSUE TITLE
[IoT Edge] ML Module deploy error

### ISSUE THREAD

**Issue by JihongL** [Azure/iotedge#1161]
Hi, I am having trouble with deploy a ml module in iot edge device.

**Environments**
- Ubuntu 18.04.2 LTS VM  on Windows 10 pro Hyper-V 2nd Gen
- iotedge : 1.0.6.1
- edge-hub : 1.0
- edge-agent : 1.0
- Moby engine : 3.0.5
- docker0 ipv4 : 172.18.0.1
- ml module image uri : lsmcprdmlws2411913023.azurecr.io/deeplearningtest:3

**iot edge config**
```
connect:
  management_uri: "http://172.18.0.1:15580"
  workload_uri: "http://172.18.0.1:15581"

listen:
  management_uri: "http://172.18.0.1:15580"
  workload_uri: "http://172.18.0.1:15581"
```

**/etc/environment**
```
export IOTEDGE_HOST="http://172.18.0.1:15580"
```

**Logs**
```
May 02 01:52:18 edge iotedged[10582]: 2019-05-02T01:52:18Z [INFO] - Pulling image lsmcprdmlws2411913023.azurecr.io/deeplearningtest:3...
May 02 01:52:40 edge iotedged[10582]: 2019-05-02T01:52:40Z [INFO] - Checking edge runtime status
May 02 01:52:40 edge iotedged[10582]: 2019-05-02T01:52:40Z [INFO] - Edge runtime is running.
May 02 01:52:42 edge iotedged[10582]: 2019-05-02T01:52:42Z [INFO] - Successfully pulled image lsmcprdmlws2411913023.azurecr.io/deeplearningtest:3
May 02 01:52:42 edge iotedged[10582]: 2019-05-02T01:52:42Z [INFO] - Creating module mlmodule...
May 02 01:52:42 edge iotedged[10582]: 2019-05-02T01:52:42Z [WARN] - Could not create module mlmodule
May 02 01:52:42 edge iotedged[10582]: 2019-05-02T01:52:42Z [WARN] -         caused by: No such image: lsmcprdmlws2411913023.azurecr.io/deeplearningtest:3

~~~~

May 02 02:14:40 edge iotedged[13434]: 2019-05-02T02:14:40Z [WARN] - Could not pull image lsmcprdmlws2411913023.azurecr.io/deeplearningtest:3
May 02 02:14:40 edge iotedged[13434]: 2019-05-02T02:14:40Z [WARN] -         caused by: Get https://lsmcprdmlws2411913023.azurecr.io/v2/deeplearningtest/manifests/3: dial tcp: lookup lsmcprdmlws2411913023.azurecr.io on 127.0.0.53:53: read udp 127.0.0.1:54766->127.0.0.53:53: i/o timeout
```

Even modify config, a ml module doesn't deployed. but other modules are deployed well.
I guess disable process identification is not applied properly. but, I follow instructions on [disable process identification](https://docs.microsoft.com/en-us/azure/iot-edge/tutorial-deploy-machine-learning#disable-process-identification) and I checked more than double.

Please let me know anything to resolve this situation.

Thanks.

**Comment by myagley** [comment]
This looks like a DNS issue. The Docker engine can't resolve the hostname `lsmcprdmlws2411913023.azurecr.io` to an IP address.

You may want to configure the Docker engine with a set of DNS servers. This configured in Docker's config file, usually located at `/etc/docker/daemon.json`. (You may need to create the file). Here is more info on the settings: https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file

An example for specifying DNS:

```json
{
    "dns": ["1.1.1.1"]
}
```

You should replace the IP address used with something that makes sense for your environment.

**Comment by JihongL** [comment]
@myagley Thanks for Response.
but, it seems didn't work.

As you said, there were no directory and file like `/etc/docker/daemon.json`, So I created that.
And input `service docker restart`, `systemctl restart iotedge`. But, mlmodule is still not deployed.
I set dns 1.1.1.1, 8.8.8.8, but both were failed.

Here is additional error messages.

**journalctl -u iotedge**
```
May 03 00:37:54 edge iotedged[91399]: 2019-05-03T00:37:54Z [INFO] - [mgmt] - - - [2019-05-03 00:37:54.508670479 UTC] "GET /modules?api-version=2018-06-28 HTTP/1.1" 200 OK 2095 "-" "-" pid(any)
May 03 00:38:02 edge iotedged[91399]: 2019-05-03T00:38:02Z [INFO] - Checking edge runtime status
May 03 00:38:02 edge iotedged[91399]: 2019-05-03T00:38:02Z [INFO] - Edge runtime is running.
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [ERR!] - Internal server error: Could not list identities
May 03 00:38:10 edge iotedged[91399]:         caused by: Could not list identities
May 03 00:38:10 edge iotedged[91399]:         caused by: Could not list modules
May 03 00:38:10 edge iotedged[91399]:         caused by: Could not perform HTTP request
May 03 00:38:10 edge iotedged[91399]:         caused by: an error occurred trying to connect: failed to lookup address information: Temporary failure in name resolution
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [INFO] - [mgmt] - - - [2019-05-03 00:38:10.021400742 UTC] "GET /identities/?api-version=2018-06-28 HTTP/1.1" 500 Internal Server Error 287 "-" "-" pid(any)
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [INFO] - [mgmt] - - - [2019-05-03 00:38:10.118465915 UTC] "GET /identities/?api-version=2018-06-28 HTTP/1.1" 200 OK 504 "-" "-" pid(any)
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [INFO] - [mgmt] - - - [2019-05-03 00:38:10.161562747 UTC] "PUT /identities/mlmodule?api-version=2018-06-28 HTTP/1.1" 200 OK 98 "-" "-" pid(any)
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [ERR!] - server connection error: (192.168.183.104:54356)
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [ERR!] - error writing a body to connection: Broken pipe (os error 32)
May 03 00:38:10 edge iotedged[91399]: 2019-05-03T00:38:10Z [INFO] - [work] - - - [2019-05-03 00:38:10.169037821 UTC] "POST /modules/%24edgeAgent/genid/636921863136718456/encrypt?api-version=2018-06-28 HTTP/1.1" 200 OK 1961 "-" "-" pid(any)
```

iotedge logs edgeAgent
```
2019-05-03 00:44:49.836 +00:00 [INF] - Executing command: "Command Group: (
  [Create module mlmodule]
  [Start module mlmodule]
)"
2019-05-03 00:44:49.840 +00:00 [INF] - Executing command: "Create module mlmodule"
2019-05-03 00:46:29.879 +00:00 [ERR] - Executing command for operation ["create"] failed.
System.Threading.Tasks.TaskCanceledException: The operation was canceled. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.GeneratedCode.EdgeletHttpClient.CreateModuleAsync(String api_version, ModuleSpec module, CancellationToken cancellationToken) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/generatedCode/EdgeletHttpClient.cs:line 270
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.Execute[T](Func`1 func, String operation)
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.CreateModuleAsync(ModuleSpec moduleSpec) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 91
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 60
2019-05-03 00:46:29.942 +00:00 [ERR] - Executing command for operation ["Command Group: (
  [Create module mlmodule]
  [Start module mlmodule]
)"] failed.
System.Threading.Tasks.TaskCanceledException: The operation was canceled. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.GeneratedCode.EdgeletHttpClient.CreateModuleAsync(String api_version, ModuleSpec module, CancellationToken cancellationToken) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/generatedCode/EdgeletHttpClient.cs:line 270
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.Execute[T](Func`1 func, String operation)
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.CreateModuleAsync(ModuleSpec moduleSpec) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 91
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 60
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 60
2019-05-03 00:46:29.959 +00:00 [ERR] - Step failed in deployment 16, continuing execution. Failure when running command Command Group: (
  [Create module mlmodule]
  [Start module mlmodule]
). Will retry in 00s.
```

**Comment by ancaantochi** [comment]
Hi @JihongL,

Do you have a DNS server on your network? Please run on your machine: `nslookup lsmcprdmlws2411913023.azurecr.io`

This command should return the nameserver that was used to resolve the domain and you can use that server to configure docker dns, example:

```
nslookup google.com
Server:  ns.someserver.com
Address:  <NS Server IP>
```

**Comment by JihongL** [comment]
I solved this issue by change Ubuntu 18.04 to Ubuntu 16.04. On Ubuntu 16.04, DNS setting was worked.
